src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,24 +96,123 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
-    def __init__(self, data_dict):
-        self._data_dict = data_dict
+def _add_project_context(
+    external_data: Dict, project_definition: ProjectDefinition
+) -> Dict:
+    """
+    Updates the external data with variables from snowflake.yml definition file.
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = {_CONTEXT_KEY: {"env": DictWithEnvironFallback({})}}
+    # If there's project definition file then resolve variables from it
+    if project_definition and project_definition.meets_version_requirement("1.1"):
+        context_data = _resolve_variables_in_project(project_definition)
+
+    context_data.update(external_data)
+    return context_data
+
+
+def _remove_ctx_env_prefix(text: str) -> str:
+    prefix = "ctx.env."
+    if text.startswith(prefix):
+        return text[len(prefix) :]
+    return text
+
+
+def string_includes_template(text: str) -> bool:
+    return bool(re.search("&{.+}", text))
+
+
+def _resolve_variables_in_project(project_definition: ProjectDefinition):
+    variables_data: DictWithEnvironFallback = cast(
+        DictWithEnvironFallback, project_definition.env
+    )
+
+    variables_with_dependencies = _check_variables_consistency(variables_data)
+    # Sort keys so we start from the shortest having lower probability of including more than one variable
+    unresolved_keys = sorted(list(variables_with_dependencies), reverse=True)
+
+    env = get_snowflake_cli_jinja_env()
+    context_data = {_CONTEXT_KEY: project_definition}
+    while unresolved_keys:
```

Comment:
how do we guarantee no infinite loop?

Reply:
`unresolved_keys` are returned from `_get_variables_with_dependencies` that makes sure
1. None of the variables depends on unknown variable (ctx.env.something) (possible infinite loop because of unknown error)
2. There's no cycles between between variables

So `unresolved_keys` is a list of variables defined in `env` that have some cross-dependencies between `env` variables. And thanks to above resolving all of them is possible.